packages/interactivity-router/src/assets/styles.ts

```diff
@@ -0,0 +1,49 @@
+export type StyleElement = HTMLLinkElement | HTMLStyleElement;
+
+const styleSheetCache = new Map< string, Promise< StyleElement >[] >();
+
+export const prepareStyles = (
+	doc: Document,
+	url: string = ( doc.location || window.location ).href
+): Promise< StyleElement >[] => {
+	if ( ! styleSheetCache.has( url ) ) {
+		const comment = window.document.createComment( url );
+		window.document.head.appendChild( comment );
+		styleSheetCache.set(
+			url,
+			[ ...doc.querySelectorAll( 'style,link[rel=stylesheet]' ) ].map(
+				( element: StyleElement ) => {
+					if ( doc === window.document ) {
+						return Promise.resolve( element );
+					}
+					if ( element instanceof HTMLStyleElement ) {
+						const cloned = element.cloneNode(
+							true
+						) as HTMLStyleElement;
+						window.document.head.appendChild( cloned );
+						cloned.sheet.disabled = true;
+						return Promise.resolve( cloned );
+					}
+					return new Promise( ( resolve, reject ) => {
+						const cloned = element.cloneNode() as StyleElement;
+						cloned.onload = () => {
+							cloned.sheet.disabled = true;
+							resolve( cloned );
+						};
```

Comment:
I'm worried about this part. That code disables the stylesheet once loaded **and** applied, so there's the risk of seeing a flash with the new styles.

In fact, I was able to reproduce this issue in Chrome with an x6 CPU slowdown, although it happens very occasionally.

I'm trying now to avoid this using `<link rel=preload as=style>` elements for the initial load. Let's see if it works.

cc: @luisherranz

Reply:
`<link rel=preload as=style>` is simply a hint; browsers don't have to follow it.

I guess it cannot be deactivated until it has been loaded, right?

```js
const cloned = element.cloneNode() as StyleElement;
cloned.sheet.disabled = true;
cloned.onload = () => resolve( cloned );
cloned.onerror = reject;
window.document.head.appendChild( cloned );
```

Maybe we can use `media`? Something like this:

```js
const cloned = element.cloneNode() as StyleElement;
const clonsed.media = 'disabled';
cloned.onload = () => {
	cloned.sheet.disabled = true;
        cloned.media = element.media;
	resolve( cloned );
};
cloned.onerror = reject;
window.document.head.appendChild( cloned );
```

Reply:
Another option we could investigate is to use `media` (`disabled` -> `all`) to control when the stylesheets are used but manually force a reflow after we've switched to `all` (like accessing `document.body.offsetHeight` for example).

Reply:
> I guess it cannot be deactivated until it has been loaded, right?

Right, because `cloned.sheet` is `null` at that moment.

> [...] manually force a reflow after we've switched to `all`

Interesting. I'll try that as well.

Reply:
Solved by https://github.com/woocommerce/gutenberg/pull/2#issuecomment-2691269596.